### PR TITLE
Handle ActionDispatch::ParamsParser::ParseError

### PR DESCRIPTION
**Pull request by Martouta**
Closes [THREESCALE-5280](https://issues.redhat.com/browse/THREESCALE-5280)

### CODE REVIEW

File: lib/three_scale/middleware/multitenant.rb [hunk 1]
```diff
@@ -111,9 +111,27 @@ def call(env)
       def _call(env)
         Thread.current[:multitenant] = TenantChecker.new(@attribute,@app,env)
         @app.call(env)
+      rescue ActionDispatch::ParamsParser::ParseError => error
```
Comment (hallelujah) [inline lib/three_scale/middleware/multitenant.rb:114]: I do not think it is the responsibility of this middleware. Rather create another one that is called just before ActionDis

It should be caught earlier in the chain, create another middleware if needed, so you will disable it on test and development but safely rescue in production

Comment (Martouta) [reply lib/three_scale/middleware/multitenant.rb:114]: > so you will disable it on test and development but safely rescue in production
I do not think that disabling it by default in `test` is a good idea 😄 We should make valid format requests in `test`, OR test that it is behaving as we desire in production when the params are not correct.

>I do not think it is the responsibility of this middleware. Rather create another one that is called just before Rack::Sendfile
> It should be caught earlier in the chain, create another middleware if needed

Gotcha 👍

Comment (hallelujah) [reply lib/three_scale/middleware/multitenant.rb:114]: Raising in development and test is better no? In production you fallback to a response.

That's how rails work with 404 and `ActiveRecord::RecordNotFound` at least. Make it configurable with sane defaults

Comment (Martouta) [reply lib/three_scale/middleware/multitenant.rb:114]: For everybody else, we spoke about this in private, and these are the conclusions:

We never raise it again for RecordNotFound and we won't raise it either for this one 😄 

https://github.com/3scale/porta/blob/7b55693bcedfed8cd707550126c1444a0187111d/app/controllers/application_controller.rb#L11
https://github.com/3scale/porta/blob/7b55693bcedfed8cd707550126c1444a0187111d/app/controllers/concerns/error_handling.rb#L8
https://github.com/3scale/porta/blob/7b55693bcedfed8cd707550126c1444a0187111d/app/controllers/concerns/error_handling.rb#L22-L24
https://github.com/3scale/porta/blob/7b55693bcedfed8cd707550126c1444a0187111d/app/controllers/concerns/error_handling.rb#L40-L65

Comment (Martouta) [inline lib/three_scale/middleware/multitenant.rb:114]: `/Users/marta/.rvm/gems/ruby-2.5.5/gems/actionpack-5.0.7.2/lib/action_dispatch/middleware/params_parser.rb`

👀

File: test/integration/admin/api/buyers_users_controller_test.rb [hunk 2]
```diff
@@ -0,0 +1,59 @@
+# frozen_string_literal: true
+
+require 'test_helper'
+
+class Admin::Api::BuyersUsersControllerTest < ActionDispatch::IntegrationTest
```
Comment (hallelujah) [inline test/integration/admin/api/buyers_users_controller_test.rb:5]: I do not think it is related to this controller

Please create a more generic one to test it ;)
Btw no need to test in controller just make a rack test is enough

Comment (Martouta) [reply test/integration/admin/api/buyers_users_controller_test.rb:5]: Awesome, will do 👍

File: test/unit/three_scale/middleware/handle_parse_error_test.rb [hunk 3]
```diff
@@ -0,0 +1,30 @@
+# frozen_string_literal: true
+
+require 'test_helper'
+
+class ThreeScale::Middleware::HandleParseErrorTest < ActionDispatch::IntegrationTest
```
Comment (hallelujah) [inline test/unit/three_scale/middleware/handle_parse_error_test.rb:5]: You do not need (but you can have ...) this integration test, as long you have the corresponding unit test
See:
https://github.com/3scale/porta/blob/3d07b9e15c22ae607dad7abc3aebca4bc0770e8a/test/unit/sidekiq_logging_middleware_test.rb
and
https://github.com/3scale/porta/blob/3d07b9e15c22ae607dad7abc3aebca4bc0770e8a/test/unit/three_scale/middleware/dev_domain_test.rb

**Pull request merged**